[MRG+1] Updates entire website design

thomasjpfan opened the pull request:
Here is a [hosted version](https://thomasjpfan.github.io/scikit-learn-website/index.html) that follows this PR.

This PR:

1. Updates the entire design of the website.
2. Keeps all features from the original website.
3. Adds custom sidebar with button on the lower left. (Mobile friendly)
4. Completely custom sphinx theme that does not inherit from `basic`. There were a bunch of js/css we did not use, this is no longer loaded.
5. There are no images in the new theme.
6. Small images were generated for the logos on the bottom of the page in the index.
7. Uses bootstrap 4.

Enjoy! :)

thomasjpfan committed:
WIP

thomasjpfan committed:
WIP

thomasjpfan committed:
WIP

thomasjpfan committed:
WIP

thomasjpfan committed:
DOC Updates website design

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into scikit-learn-bootstrap

thomasjpfan committed:
REV

thomasjpfan committed:
STY Fix

thomasjpfan committed:
TST [doc build]

thomasjpfan committed:
REV No logos

thomasjpfan committed:
TST [doc build]

thomasjpfan committed:
DOC Adjust z-index of nav

thomasjpfan committed:
ENH Mobile first search [doc build]

jnothman commented on doc/templates/documentation.html:
```diff
@@ -0,0 +1,105 @@
+    </div>
+    <div class="col-md-4 mb-2">
+      <div class="card h-100 sk-documentation-index-card mx-md-2 mb-2">
+        <a href="#" class="stretched-link sk-documentation-index-anchor"></a><h4 class="card-header">Quick Start</h4>
```
Wrong title

jnothman commented:
Wow! Loving it. I like the "minimal change to get it working" approach

rth commented:
Very impressive @thomasjpfan !

> Completely custom sphinx theme

I think overall this is great, though if we can afford a professional web-designer (cc @amueller) at least to review this and maybe propose some improvements on top this PR it would be ideal.  Personally, I can say that the rendering looks good for me, but I really don't know to review it (with respect to cross-browser compatibility, best web-design practices etc). @cmarmo was also working on improving and if there is a way to collaborate on it somehow it would be great.

A side note, do we want to bundle minified jquery and bootstrap (as opposed to using a CDN)? I get they this allows to use the doc offline, but they are not that small and a prone to be updated frequently.

rth commented:
For instance, I also wonder if bootstrap is still the best way to go in the frontend world. Though I agree it's a clear improvement over our current website in any case.

rth commented:
> For instance, I also wonder if bootstrap is still the best way to go in the frontend world. Though I agree it's a clear improvement over our current website in any case.

FWIW, pytoch also uses boostrap but their theme is very customized.

thomasjpfan committed:
BUG Fixes

thomasjpfan committed:
BUG Style fixes

thomasjpfan committed:
ENH Does not load mathjax in index

thomasjpfan committed:
ENH Whatsnew badges

thomasjpfan committed:
TST [doc build]

thomasjpfan committed:
ENH Enhances image display [doc build]

thomasjpfan committed:
ENH Fixes headerlink jumping  [doc build]

thomasjpfan committed:
ENH Increases constrast between sidebar and main [doc build]

thomasjpfan committed:
ENH Table style fix [doc build]

thomasjpfan commented:
> Personally, I can say that the rendering looks good for me, but I really don't know to review it

Not having something like webpack to do auto prefixing is a little rough. I have been developing in firefox/chrome/safari. (edge is moving to chromium) I will get my hands on IE and see what it looks like.

> FWIW, pytorch also uses bootstrap but their theme is very customized.

They vendor their bootstrap as well and not use the bootstrap CDN. (It is pretty nice to view docs locally)

> For instance, I also wonder if bootstrap is still the best way to go in the frontend world. Though I agree it's a clear improvement over our current website in any case.

Bootstrap is really nice to prototype design in because I was most familiar with it. Now that we have a design, I think it is possible to use something smaller, like purecss. We would only need custom js for the "who uses scikit-learn" carousel and maybe for the navbar folding in mobile. Although, using a familiar/popular css framework like Bootstrap makes it easier for others to contribute/modify.

thomasjpfan committed:
ENH Adds back links to other versions [doc build]

thomasjpfan committed:
ENH Padding to search results [doc build]

TomDLT commented:
[rendered website](https://72179-843222-gh.circle-artifacts.com/0/doc/index.html)

thomasjpfan committed:
ENH Color fixes [doc build]

thomasjpfan committed:
ENH CSS sure is fun [doc build]

thomasjpfan committed:
ENH CSS sure is fun [doc build]

thomasjpfan committed:
ENH UX is fun [doc build]

thomasjpfan committed:
ENH UX is fun [doc build]

thomasjpfan committed:
ENH Adds table of contents page [doc build]

thomasjpfan committed:
ENH Mobile title fix [doc build]

thomasjpfan committed:
ENH Link to anchor fix [doc build]

thomasjpfan committed:
ENH JS comments [doc build]

thomasjpfan committed:
ENH UX and css is fun [doc build]

thomasjpfan committed:
BUG Fixes broken link [doc build]

thomasjpfan committed:
ENH UX is awesome [doc build]

thomasjpfan committed:
ENH UX is awesome [doc build]

thomasjpfan committed:
ENH UX is awesome [doc build]

thomasjpfan committed:
ENH Adds border to the right [doc build]

thomasjpfan committed:
ENH Better anchor links support [doc build]

thomasjpfan committed:
ENH Smaller header size [doc build]

thomasjpfan committed:
ENH Smarter anchor links [doc build]

thomasjpfan committed:
BUG Unhides navbar  [doc build]

thomasjpfan committed:
ENH Auto hiding navbar [doc build]

thomasjpfan committed:
ENH Nicer looking google search [doc build]

thomasjpfan committed:
ENH Nicer looking landing [doc build]

thomasjpfan committed:
BUG Nowrap for code [doc build]

thomasjpfan commented:
Here is a hosted version that follows this PR: https://thomasjpfan.github.io/scikit-learn-website/index.htmla

thomasjpfan committed:
ENH Hide navbar if hashtarget is on top first load

thomasjpfan committed:
ENH Better navhide support for mobile [doc build]

thomasjpfan committed:
ENH Adds margins [doc build]

thomasjpfan committed:
ENH Nicer google search [doc build]

thomasjpfan committed:
CLN Uses sk- naming convention [doc build]

jnothman commented:
Not sure where we should start reviewing here. Will the new
versionwarning.js need adjusting?

thomasjpfan commented:
`versionwarning.js` should not need adjusting. The new layout has an element with `class="body"` and `role="main"` to place the alert.

<img width="1220" alt="Screen Shot 2019-09-01 at 7 36 38 AM" src="https://user-images.githubusercontent.com/5402633/64075850-42845500-cc8b-11e9-8f9b-f22b527d7602.png">

thomasjpfan committed:
ENH Dynamicaly adjusts navbar [doc build]

thomasjpfan commented:
> Not sure where we should start reviewing here

Since most of this PR is visual, we most likely need time to use the site. Most of the features are quality of life improvements:

1. The sidebar now scrolls independently from the main page.
2. The navbar now hides when a user scrolls down into a page and comes back when a user scrolls up. (This is where most of the new javascript lies)
3. Google search now displays results with `position: fixed` which makes it more usable and will open in the current window (not in a new tab).
4. All javascript is moved to the end right before the body so the javascript runs at the end.
5. Mathjax is not loaded in `documentation` and `index`, this reduces the page size and load speed.

From a doc maintenance point of view: 
1. The orphaned `index.rst` and `documentation.rst`, files are moved to `templates` as HTML files, giving us full control of its layout. (There were `p` and other unneeded html elements in the `rst` version)
2. The theme does not inherit from `basic` so the layout is completely described in `layout.html`.
3. If we want to add sitewide javascript, we add it into `javascript.html`. This adds to the HTML slightly, but removes GET requests to javascript files.
4. There is a new `contents.rst` that holds the global TOC.

Here are some performance results (with cache disabled):

1.  The [new index page page](https://thomasjpfan.github.io/scikit-learn-website/index.html) makes 39 requests and loads in 600ms, compared to 58 requests and loads in 1s.
2. A huge page such as [the new Generalized Linear Models page](https://thomasjpfan.github.io/scikit-learn-website/modules/linear_model.html) makes 58 requests compared to 69 requests on master, which decreases load time by 12%.
3. A huge page with just text such as the [new API reference page](https://thomasjpfan.github.io/scikit-learn-website/modules/classes.html) makes 25 requests compared to 34, which decreases load time by 25%.

thomasjpfan committed:
ENH Locally scopes new javascript [doc build]

jnothman commented on doc/contents.rst:
```diff
@@ -0,0 +1,16 @@
+=================
+
+.. Define an order for the Table of Contents:
+
```
Does this want `big_toc_css` for consistency?

jnothman reviewed (commented):
Wow!

While you're adding contents.rst... do you mind looking at the need for and structure of [`preface.rst`](https://thomasjpfan.github.io/scikit-learn-website/preface.html)? Its "welcome to scikit-learn" heading is awkward and hides its contents from [`contents.html`](https://thomasjpfan.github.io/scikit-learn-website/contents.html). I'm also not sure if the comment regarding the top-level heading there for TeX is justified any longer.

jnothman reviewed (commented):
Something else that I've noticed is that the release notes don't seem to be linked to from anywhere prominent (other than news). Wondering how to best address that.

thomasjpfan committed:
CLN Media with screen [doc build]

thomasjpfan committed:
CLN Nicer google search box [doc build]

jnothman commented on doc/templates/documentation.html:
```diff
@@ -0,0 +1,113 @@
+      <div class="card h-100 sk-documentation-index-card mx-md-2 mb-2">
+        <h4 class="card-header">Other Versions</h4>
+        <div class="card-body p-3">
+          <p class="card-text">
```
Don't wrap a `<ul>` in a `<p>`. The browser will automatically close the `<p>` and so insert some visual padding north of the list.

(Perhaps do some HTML validation)

thomasjpfan replied:
I'll do more HTML validation on more pages to make sure everything is okay.

jnothman commented on doc/themes/scikit-learn-modern/nav.html:
```diff
@@ -0,0 +1,52 @@
+      <div class="sk-search-form">
+          <div class="gcse-search" id="cse" data-linktarget="_parent"></div>
+      </div>
+      </form>
```
stray closing tag. (Found by https://validator.w3.org/nu/?doc=https%3A%2F%2Fthomasjpfan.github.io%2Fscikit-learn-website%2Fdocumentation.html)

thomasjpfan replied:
Ah yes, thank you for catching this! This was left over from a custom search bar that was converted into a google search bar.

jnothman commented:
I like the funky interaction between the menu and the navbar!

jnothman commented:
I note that "Toggle menu" appears but is unusable in a browser with Javascript disabled.

jnothman commented:
Actually, javascript wasn't disabled, just external JS sources weren't available.

thomasjpfan commented:
> Actually, javascript wasn't disabled, just external JS sources weren't available.

Makes sense, I'l rewrite my additional javascript to not depend on jquery.

jnothman commented:
> Makes sense, I'l rewrite my additional javascript to not depend on jquery.

I didn't mean external as in internet. I think it's fine to depend on jQuery especially if it's a standard CDN that every browser already has cached.

I meant that I removed access to any js files.

You can also use css for the toggle: make the toggle button the label for a checkbox whose state can be used in the CSS selector for the menu (e.g. the checkbox is a sibling of the menu). Then `#menu-toggle + #menu { display: none } #menu-toggle:checked + #menu { display: block }` or whatever... I'm sure you can find blogs about this hack

thomasjpfan committed:
ENH Pure css sidebar toggling

thomasjpfan committed:
BUG Fix on html [doc build]

thomasjpfan commented:
PR updated with a pure CSS solution on the sidebar toggle. Thank you for the tip @jnothman ! :)

thomasjpfan commented:
@rth 

As for [purecss](https://purecss.io), I think if we vender or remove the carousel on bottom right in the index page, this PR can use [purecss](https://purecss.io).

thomasjpfan committed:
DOC Better padding [doc build]

thomasjpfan committed:
ENH Better style in documentation [doc build]

thomasjpfan committed:
ENH Better image style in index [doc style]

thomasjpfan committed:
DOC Better padding [doc build]

thomasjpfan committed:
DOC Fix testimonials [doc build]

thomasjpfan commented:
After trying to get this to use purecss, I observed:

1. purecss uses 19kb less space when compared to bootstrap (compressed + minified)
2. adapting purecss would require more custom css and javascript to account for missing components such as a bootstrap's card, navbar, and bootstraps nice layout utilities.

The benefit of purecss is that it takes up less space uncompressed in the repo. The negative is that it gives us more code to maintain. Overall I am -0.5 on using purecss.

thomasjpfan commented:
> Something else that I've noticed is that the release notes don't seem to be linked to from anywhere prominent (other than news). Wondering how to best address that.

Hmm wWhat about having a link on the top as well:

<img width="604" alt="Screen Shot 2019-09-02 at 1 39 04 PM" src="https://user-images.githubusercontent.com/5402633/64128960-13eba480-cd87-11e9-91ad-0d14141f0e57.png">

thomasjpfan committed:
ENH Adds whats new button on top [doc build]

jnothman commented:
I like how that puts the current version front and centre

jnothman commented:
I like less how the maintainer needs to change it :/

thomasjpfan committed:
BUG Properly adds big_toc [doc build]

GaelVaroquaux commented:
Overall, I like the new look. A few quick comments:

* On the front page:
  - I think that the logo should be larger (also true of other pages, but less important)
  - The colors of the gradient could be made a bit lighter
  - To decrease horizontal scaling, it would be nice if the "whatsnew" could be on the right for large screens (which might raise the need of adapting its colors). Other option: fold the black bullet points to the right.
  - We might need a bit more white space between the block (the images can be made a tiny bit smaller, if it helps)

* On the top bar:
  - I would prefer more contrast by default. When viewing in bright sunshine, less contrast means more difficulties to read
  - I miss the drop-down menus, which were useful to get somewhere with few clicks

* On the examples page, the popover of examples does not overflow as it should on the right of the page (I think that it is related to one of the properties set by bootstrap on the enclosing divs, but I couldn't find it even after playing a bit).

* Sidebar:
  - I personally do not mind if we remove the "toggle menu" button, and replace it by a well-chosen "media" selector, so that the menu vanishes automatically on narrow screens.
  - It would probably make doc pages easier to read if their was a bit more white space between the sidebar and the text on large screen (in general, white space tends to improve readability)
  - (difficult one :D): We have too many things in our sidebar. As a result it overflows vertically a lot and forces scrolling on many pages. It might be good on very large screens to have a sidebar on the left and one on the right. Also: "div.sk-sidebar-toc-wrapper ul ul ul { font-size: small;}"

* I think that we could use a bit more vertical space above h1, h2, and maybe h3

* Nitpick: more whitespace between blocks on https://thomasjpfan.github.io/scikit-learn-website/documentation.html

* Asking for the moon: can we highlight the current subsection as we scroll in the sidebar :).

I've already asked for a lot :). So I'll stop there, unless I see other visual bugs.

thomasjpfan committed:
DOC Build pdf correctly [doc build]

thomasjpfan committed:
BUG Fix [doc build]

thomasjpfan committed:
CLN Remove a header

thomasjpfan commented:
> While you're adding contents.rst... do you mind looking at the need for and structure of

@jnothman The preface makes the PDF table of contents a bit nicer, since it combines the documents in the preface into a single top level section. As for the comment, about "top level heading needed for LaTeX TOC in sphinx<=1.3.1", was removed along with the header. This has the added benefit of breaking the TOC into subsection in the global toc in the pdf:

<img width="478" alt="Screen Shot 2019-09-03 at 12 02 05 AM" src="https://user-images.githubusercontent.com/5402633/64143454-17f6e100-cdde-11e9-997c-d4933560a6c8.png">

If we want to hide these sections from the global toc, we can revert back to using two headers in `preface.rst`.

As for the "welcome to scikit-learn" phrasing, it kind of fits in this book-like format.

thomasjpfan committed:
CLN Fix overflow

amueller commented:
hm I thought we'd get a designer? :P

thomasjpfan commented:
We can still get a UX designer. :) (I should be able to persuade sphinx to do what a designer wants)

@GaelVaroquaux I agree with most of your points.

> To decrease horizontal scaling, it would be nice if the "whatsnew" could be on the right for large screens (which might raise the need of adapting its colors). Other option: fold the black bullet points to the right.

If we make this part too small, it would not "stand out as much". For example: https://pytorch.org/, https://jupyter.org/, https://www.tensorflow.org/ have huge  "jumbotrons". Although, I agree that moving the black bullets to the right works.

> I miss the drop-down menus, which were useful to get somewhere with few clicks

It is useful. Although we would need to rethink where to put the "documentation" page. Having just the little arrow be the activator for the dropdown is not mobile friendly. I think if we want to keep this dropdown, it would be good to promote some items into the navbar, such as the user guide and API. The rest can be in the dropdown. (Remove home since the logo goes home)

> On the examples page, the popover of examples does not overflow as it should on the right of the page (I think that it is related to one of the properties set by bootstrap on the enclosing divs, but I couldn't find it even after playing a bit).

I am happy you are playing with the design. I am looking into this.

>  We have too many things in our sidebar. As a result it overflows vertically a lot and forces scrolling on many pages. It might be good on very large screens to have a sidebar on the left and one on the right. 

This is an [option pytorch has explored](https://pytorch.org/tutorials/beginner/nn_tutorial.html). I personally have never used the right sidebar. I feel like it makes the page "too busy".

> Asking for the moon: can we highlight the current subsection as we scroll in the sidebar :).

This will be fun.

@amueller There are several places that can improve in navigation. Most of this is inspired by https://pytorch.org/, since they have a significant amount of content as well. The biggest difference is that we have a huge user guide.

1. The user guide is basically a book. It would be good to use a UX designed for books such as: https://christophm.github.io/interpretable-ml-book/
2. As for people searching for things, the google search on the top right has been redone. It now feels like it is part of the website and opens links on the same page.
3. The user guide and API should most likely be in the navbar.

GaelVaroquaux commented:
> We can still get a UX designer. :) (I should be able to persuade sphinx to do what a designer wants)

How about we merge your iteration first? They are a net improvement.

> I agree that moving the black bullets to the right works.

Let's do this!

> I think if we want to keep this dropdown, it would be good to promote some items into the navbar, such as the user guide and API. The rest can be in the dropdown. (Remove home since the logo goes home)

Let's do this too!

> >     On the examples page, the popover of examples does not overflow as it should on the right of the page

> I am happy you are playing with the design. I am looking into this.

I believe that you fixed that.

> >     Asking for the moon: can we highlight the current subsection as we scroll in the sidebar :).

> This will be fun.

I think that bootstrap has a way to do this. Elsewhere, I have javascript
that does this, but if we go bootstrap, we might as well go bootsrap all
the way.

>  1. The user guide is basically a book.

Agreed.

> It would be good to use a UX designed for books such as: https://christophm.github.io/interpretable-ml-book/

I find this design a bit sad. A while ago, I worked on the design of the scipy lecture
notes to make them more book-like, while keeping branding:
http://scipy-lectures.org/packages/scikit-learn/index.html

Overall, I think that it's about more white and white space.

>  3. The user guide and API should most likely be in the navbar.

+1

jorisvandenbossche commented:
> > > Asking for the moon: can we highlight the current subsection as we scroll in the sidebar :).
> > >
> > This will be fun.
> >
> I think that bootstrap has a way to do this

Yes, in bootstrap it can be rather easily done with scrollspy. 
I am working with someone on new sphinx theme for the pandas docs as well, also based on bootstrap, and there our right side bar highlights where we are on the page. See eg https://dev.pandas.io/pandas-sphinx-theme/pr-bootstrap-sphinx-theme/user_guide/reshaping.html (note we only focused on structure and functionality for now, not yet styling of the details, the content div is also still too wide)

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into scikit-learn-bootstrap

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
CLN Displays depth 2 on sidebar

GaelVaroquaux commented:
You've removed the flashy gradient in the latest version of the front page. I player a bit with 
https://cssgradient.io/ and here is a gradient that I find works well (goes on div.sk-landing-bg)
```
background: linear-gradient(90deg, rgba(42,98,128,1) 0%, rgba(52,153,205,1) 17%, rgba(255,243,211,1) 59%, rgba(255,178,96,1) 100%);
}
```

Resulting frontpage:
![image](https://user-images.githubusercontent.com/208217/64480383-949a0e80-d194-11e9-8504-01ac187d9ccf.png)

GaelVaroquaux commented:
Other small comment: I was confused that "User Guide" was not found in the "Docs" menu, and then I realized that it is because it's already on the top menu. Maybe rename "Docs" to "More docs"?

thomasjpfan committed:
DOC Make gradients great again

thomasjpfan commented:
Ohh thats great! I angled the gradient to the lower right corner so it looks up on mobile as well:

<img width="504" alt="Screen Shot 2019-09-07 at 6 04 29 PM" src="https://user-images.githubusercontent.com/5402633/64480691-28220e00-d19a-11e9-8eda-1f265243a1a0.png">

thomasjpfan committed:
DOC Update

thomasjpfan committed:
DOC More tab

thomasjpfan committed:
DOC Adjusts padding

thomasjpfan committed:
DOC Adjusts margins

thomasjpfan committed:
DOC Fixes choose the right estimator

thomasjpfan committed:
DOC Dynamic versioning

thomasjpfan committed:
CLN Cleaner relinks

thomasjpfan committed:
CLN Ligher background hover

GaelVaroquaux commented:
I made a small PR with suggestions adding horizontal whitespace on large screens: https://github.com/thomasjpfan/scikit-learn/pull/9

adrinjalali commented:
I really like how it looks. I think @NicolasHug mentioned he wanted to refresh the quick start guide. Right now the quick start is only visible through the "more" menu, but I guess it should be the one people see first.

NicolasHug commented:
Yes! getting-started PR is here https://github.com/scikit-learn/scikit-learn/pull/14920

I think it'd make sense to have it  as a button in the headline (maybe instead of having the whatsnew?), like e.g. [pytorch](https://pytorch.org/)

GaelVaroquaux committed:
STY more horizontal spacing on index (#9)

thomasjpfan committed:
CLN MORE WHITESPACE

thomasjpfan committed:
DOC Rename index

thomasjpfan committed:
DOC More natural media

thomasjpfan committed:
DOC Change breakpoint

thomasjpfan committed:
DOC Fix CSS

GaelVaroquaux commented:
On my computer, hovering over the links makes their color change (from cyan-like to a deeper blue, which looks a bit like a browser default).

This seems to me like a css bug, no?

GaelVaroquaux commented:
Other small detail (sorry, I am reporting them as I see them): the examples used to be numbered (via the numbering of the toc), it's no longer the case. I found that useful while teaching.

GaelVaroquaux commented:
Also, it seems that the rendering bug of overlay of the popovers in the example gallery is back.

thomasjpfan committed:
DOC Update quick start wording

thomasjpfan committed:
DOC Slightly lesss dark button

thomasjpfan committed:
CLN Address comments

thomasjpfan commented:
> (sorry, I am reporting them as I see them):

No problem! Since these are all visual, I suspect this would be the case.

> This seems to me like a css bug, no?

This is on purpose and inspired by https://getbootstrap.com/docs/4.3/components/buttons/ I adjusted some colors a little.

> Also, it seems that the rendering bug of overlay of the popovers in the example gallery is back.

This was a design choice to try to keep child containers inside its parents container space. `overflow-x: scroll` was set because of how huge tables in the user guide will cause the whole page to scroll left and right. This is a bit annoying on mobile. As for this issue, I adjusted `overflow-x` to be visible in the examples to allow for the popovers to be visible.

> Other small detail (sorry, I am reporting them as I see them): the examples used to be numbered (via the numbering of the toc), it's no longer the case. I found that useful while teaching.

The examples are not currently numbered on master: https://scikit-learn.org/stable/auto_examples/index.html I am okay with getting this numbered as long as it does not number everything else. (The API reference looks extremely busy numbered)

jnothman commented:
Let's not leave this open too long... rather merge and road test..?

thomasjpfan committed:
CLN Improves index image placement

GaelVaroquaux commented:
> The examples are not currently numbered on master: https://scikit-learn.org/stable/auto_examples/index.html

Sorry, I got mixed up with nilearn.

> I am okay with getting this numbered as long as it does not number everything else. (The API reference looks extremely busy numbered)

+1

thomasjpfan committed:
ENH Numbered examples

GaelVaroquaux commented:
I made an extra minor enhancement: https://github.com/thomasjpfan/scikit-learn/pull/10

But I agree with @jnothman : this is good enough. It's a net improvement over what we have, and there is no obvious functionality loss or new bugs. +1 for merge on my side.

GaelVaroquaux reviewed (approved):
LGTM. +1 for merge on my side. We can test-run and improve from this, if needed.

NicolasHug commented:
I haven't reviewed the code, only the design, and I really like it too. +1 for merging

thomasjpfan committed:
DOC Improves navigation for index toctree files

thomasjpfan committed:
DOC Slightlu smaller more info landing

GaelVaroquaux committed:
STY: example download note to the right (#10)

thomasjpfan committed:
DOC Places example link next to content

thomasjpfan committed:
STY Image fix

amueller commented:
ok does that mean we decided against hiring a designer? There's several issues on my browser with the current hosted version...

thomasjpfan committed:
DOC Fix

GaelVaroquaux commented:
> ok does that mean we decided against hiring a designer?

One danger with hiring a designer is having to translate the resulting
work to our toolchain. Unless the person works directly with it.

As far as I am concerned, the current result is good enough. Any chance
that Thomas can fix the resulting problems?

amueller commented:
> One danger with hiring a designer is having to translate the resulting
work to our toolchain. Unless the person works directly with it.

yeah I would only hire someone that works directly in sphinx.

jnothman commented:
If you can define a spec for how a designer should improve on this then we
can go ahead and do that

thomasjpfan committed:
DOC Better table handling

thomasjpfan committed:
Bug Fix scrolling

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into scikit-learn-bootstrap

thomasjpfan committed:
DOC Fix global toc

GaelVaroquaux commented:
The download box of examples now is partly maxed by the top bar.

This can be fixed by adding  "top: 50px;" to the corresponding css rules.

GaelVaroquaux commented:
> The download box of examples now is partly maxed by the top bar.

It seems that it's happening only on some examples. I'm not sure why. For instance here: https://thomasjpfan.github.io/scikit-learn-website/auto_examples/ensemble/plot_adaboost_regression.html#decision-tree-regression-with-adaboost

thomasjpfan committed:
BUG Fix header example linking

thomasjpfan commented:
https://thomasjpfan.github.io/scikit-learn-website/auto_examples/ensemble/plot_adaboost_regression.html has no anchor which goes to the top of the page, including the download bar.

https://thomasjpfan.github.io/scikit-learn-website/auto_examples/ensemble/plot_adaboost_regression.html#decision-tree-regression-with-adaboost contains the anchor which places the header at the very top. I updated this PR with a `:before` css method to give some space for the download bar to show when the header anchor is linked.

GaelVaroquaux commented:
Right, but it seems to me that now you really need the "top" rule: all examples are now consistent with the same behavior that has overlap. That behavior is not desirable, in my eyes.

thomasjpfan commented:
We have this behavior on master:

https://scikit-learn.org/stable/auto_examples/ensemble/plot_adaboost_regression.html#sphx-glr-auto-examples-ensemble-plot-adaboost-regression-py

https://scikit-learn.org/stable/auto_examples/ensemble/plot_adaboost_regression.html

(Although I'll try to get the desired behavior on this PR)

GaelVaroquaux commented:
I am talking about the div linking to the download and binder, which now hangs on the right (and doesn't in master). I think that it needs "top: 50px;" to avoid overlapping with the top bar.

thomasjpfan committed:
STY Adds top

thomasjpfan commented:
Sorry I misunderstood. I see now you were referring to the div when the screen is extra large.

Update PR to add the `top`.

GaelVaroquaux commented:
Yes, I had realized that we were talking across each other :).

Indeed, your latest iteration fixes the problem.

GaelVaroquaux commented:
Small detail (not a blocker): I find it strange that the fonts of the top bar are smaller on the home page than on other pages.

thomasjpfan committed:
STY Fix navbar font

jaquesgrobler commented:
This looks really good -
Just a note or two:

on iPhoneX, for the following page: https://thomasjpfan.github.io/scikit-learn-website/developers/contributing.html#testing-and-improving-test-coverage
The content isn't filling the viewport - leaving a right gutter visible:

Another thing i noted, 
On desktop view, for https://thomasjpfan.github.io/scikit-learn-website/developers/contributing.html#testing-and-improving-test-coverage
, one is able to expand some of the line items by clicking the arrow..
This doesn't, however, work for the mobile version - clicking the arrow will open the parent link immediately.
![expand](https://user-images.githubusercontent.com/1378870/64587607-33468b00-d3a0-11e9-8821-a5cb76febbe4.gif)

Looks great though! Good job :)

jaquesgrobler commented:
On https://thomasjpfan.github.io/scikit-learn-website/developers/contributing.html#ways-to-contribute
for Mobile view, the hamburger menu is missing at the top.
Upon clicking **Toggle Menu** - which is only visible if you scroll down a wee bit, the sidemenu content will open without the content fixed, appearing to be an empty sidemenu until you scroll up:

![expand2](https://user-images.githubusercontent.com/1378870/64588041-840ab380-d3a1-11e9-9972-0312164d8d9d.gif)

jaquesgrobler commented:
Not an issue this, but something to perhaps consider (if mobile view is important)
For the examples page https://thomasjpfan.github.io/scikit-learn-website/auto_examples/index.html
Thumbnails are left aligned in mobile view.
They could be centered with a simple flex on the section container IMHO.

Something like this:
![centerthumbnails](https://user-images.githubusercontent.com/1378870/64588400-9d602f80-d3a2-11e9-83c9-66a9ed4b5bfb.gif)

thomasjpfan committed:
BUG Fixes minior bug scrolling for mobile

thomasjpfan commented:
@jaquesgrobler 

> Another thing i noted,
On desktop view, for https://thomasjpfan.github.io/scikit-learn-website/developers/contributing.html#testing-and-improving-test-coverage
, one is able to expand some of the line items by clicking the arrow..
This doesn't, however, work for the mobile version - clicking the arrow will open the parent link immediately.

This was inherited from the original version of this TOC page. I am going to rethink these long TOC  pages in terms of functionality.

> Upon clicking Toggle Menu - which is only visible if you scroll down a wee bit, the sidemenu content will open without the content fixed, appearing to be an empty sidemenu until you scroll up:

I am failing to reproduce that state. (There seems to be strange things going on with chrome that I am not getting with the other browsers which may be causing this issue)

> for Mobile view, the hamburger menu is missing at the top.

When linking to an anchor, the anchor goes to the very top of the page. There is a `:before` css hack to show the header for most cases I am thinking of implementing.

> Not an issue this, but something to perhaps consider (if mobile view is important)

This whole PR was written with mobile first and @media were added for higher widths, so it is important. Sphinx galley's CSS is written in such a way that the thumbnails are floated to the right, so the popover can be seen when hovering over it. The center is okay on mobile, but it would need to be removed when the screen gets bigger.

thomasjpfan commented:
> Upon clicking Toggle Menu - which is only visible if you scroll down a wee bit, the sidemenu content will open without the content fixed, appearing to be an empty sidemenu until you scroll up:

There is very weird behavior when using a browser that is pretending to be a phone, that does not happen on an actual phone.

thomasjpfan commented:
> This doesn't, however, work for the mobile version - clicking the arrow will open the parent link immediately.

Just tried this on a phone and touching an arrow will expand the TOC.

thomasjpfan commented:
BTW the arrow is a pretty bad touch target.

thomasjpfan committed:
STY Makes TOC arrow more touch friendly

thomasjpfan committed:
BUG Nicer header behavior

thomasjpfan committed:
CLN Remvoe debugging prints

thomasjpfan committed:
STY Slightly lower margin-top

thomasjpfan committed:
DOC Show note in examples

thomasjpfan committed:
STY Nicer touch target for TOC

thomasjpfan committed:
STY Better scroll

GaelVaroquaux commented:
I have the feeling that white space around the border of the "out" yellow divs, in the examples, has appeared recently. I personally do not really like that.

thomasjpfan committed:
STY Better sidebar scrollbar placement

thomasjpfan committed:
STY Fix output

thomasjpfan committed:
DOC Fix download note

thomasjpfan committed:
REV Examples depth

thomasjpfan commented:
> I have the feeling that white space around the border of the "out" yellow divs, in the examples, has appeared recently. I personally do not really like that.

1. Fixed.

2. As for the touch targets in a TOC with arrows to expand and collapse, they are now a little bigger to allow for better touch support.

3. The visual bug with the sidebar that @amueller found is now fixed. (CSS hacking)

4. Links from the examples page to the documentation now shows the the download note when navigating from the examples gallery index page.

Im currently thinking of refining the navigation. For example, if one is on the `ExtraTreesClassifier` class doc, it would be nice for the sidebar show the other classes in the `tree` module.

lesteve commented:
> Completely custom sphinx theme that does not inherit from basic. There were a bunch of js/css we did not use, this is no longer loaded.

Just a question on this: is there a cost in inheriting from the `basic` theme? My feeling is that not inheriting from `basic` can have some unexpected consequences. It seems like sphinx sometimes fixes things in the `basic` theme with the assumption that themes inherit from `basic` (I am guessing all sphinx themes included by sphinx, e.g. `alabaster` do), for example [this](https://github.com/sphinx-doc/sphinx/pull/5976).

I came across the case recently where we were using a theme based on ReadTheDocs theme (`sphinx_rtd_theme`) which was not inheriting from sphinx `basic` which causes some troubles, see https://github.com/dask/dask-jobqueue/pull/301 for more details. The simplest solution I could find was to include the CSS from the `basic` theme in `conf.py`.

This was not trivial to debug because:
1. I am not a sphinx expert at all (I expect this is the case for other core devs) and I very quickly was out of my depth. For example, I had no idea that sphinx had a HTML5 writer and a HTML4 writer.
2. I am not an HTML/CSS expert at all. It did not occur to me that a `:` separator could be done with CSS only.
2. it took me a while to figure out at which level the fix should be done: numpydoc, sphinx, sphinx theme, or others.

thomasjpfan commented:
@lesteve 

We ran into a similar issue on master when upgrading to sphinx 2 and master inherits from basic. This was because of how sphinx 2 changed the HTML structure for API docs. The fix was about +/- 15 lines of CSS.

Since we pin our version of sphinx:

https://github.com/scikit-learn/scikit-learn/blob/f24a501e0645713e9691626469f280930e120c00/build_tools/circle/build_doc.sh#L131

so we should not run into issues until we upgrade sphinx to the next major version. (I have not experienced issues going from 1.6->1.8)

lesteve commented:
I am not sure I fully understand your answer. Are you saying that inheriting from `basic` is not enough to avoid the two problems mentioned (one in dask-jobqueue the other in scikit-learn) above? In dask-jobqueue case, adding sphinx `basic.css` was the simplest fix.

Full disclosure: I don't really undersand how sphinx themes work and whether when you inherit from a theme you get its CSS too.

thomasjpfan commented:
In our case, inheriting from `basic.css` was not enough. If `basic.css` is included in this PR, I will still expect upgrading to Sphinx 3.x will cause us to adjust some of our CSS.

As for @amueller’s comment and having worked on that PR, I had to add more CSS to undo changes  Sphinx’s 2.x made to `basic.css` to match the style we had in our original website. `basic.css` made it harder to maintain. (It is especially bad when bootstrap.css and basic.css do not play well together)

This PR only has two css files, bootstrap and our own. Our CSS is designed to work well together with bootstrap. 

Regarding your disclosure: Inheriting from a theme gets its CSS and it’s JS. (Sphinx templating is nuts, prepare to dive into their repo to figure out what is going on 😅)

lesteve commented:
OK I'll take your word for it if you think not using `basic.css` is more maintainable. Thanks for all the details!

lesteve commented:
A few things I noticed while browsing some random pages:
* on the home page the images in "Classification", "Regression" etc ... have different sizes and vertical + horizontal centering which creates a "something is a bit off" feeling. I am not sure this is still true in your last commit so I pushed a commit with `[doc build]`. In case the problem is still here, I am not sure what level it is better to tackle this (CSS, examples generating the images, other)
* Weird scrollbar-like UI elements e.g. at the bottom of [this page](https://73844-843222-gh.circle-artifacts.com/0/doc/install.html#third-party-distributions)
![image](https://user-images.githubusercontent.com/1680079/64757133-68361780-d531-11e9-816f-29532415af39.png)
* On the home page ([CircleCI build](https://73844-843222-gh.circle-artifacts.com/0/doc/index.html)), button don't have the same color (class is `sk-landing-btn` for "Quick Start" and others at the top of the page and `btn-warning` for "help us, donate" and "Cite us" at the bottom) 
![image](https://user-images.githubusercontent.com/1680079/64757325-16da5800-d532-11e9-8a01-e25d9f8cc75a.png)
![image](https://user-images.githubusercontent.com/1680079/64757311-0a55ff80-d532-11e9-8168-fb0e27b3fd3d.png)

lesteve committed:
Trigger full CircleCI build [doc build]

jnothman commented:
Minor comment on the homepage: I find the bullet points on the right have become more prominent than "machine learning in Python". This is not true in the incumbent design since "machine learning in python" is essentially a heading to the bullets. Make "ML in Python" more prominent by increasing letter-spacing or font-size or both. It gets more prominence if the letter-spacing makes the slogan close to the same width as "scikit-learn":
![image](https://user-images.githubusercontent.com/78827/64759399-4f058900-d57a-11e9-8243-a7cb73f91c69.png)

rather than

![image](https://user-images.githubusercontent.com/78827/64759407-59278780-d57a-11e9-9a52-7dc6cedc2f70.png)

thomasjpfan committed:
BLD scroll bars in section [doc build]

thomasjpfan committed:
BLD Addresses @lesteve comments

thomasjpfan committed:
CLN Address @jnothman comments

ogrisel commented:
Here is the https://web.dev/measure report:

https://lighthouse-dot-webdotdevsite.appspot.com/lh/html?url=https://thomasjpfan.github.io/scikit-learn-website/index.html

Overall it looks not to bad.

In particular, some text / background contrast could be improved (the blue of the examples button could be darker).

There are missing alt text for the example images used to illustrate the main documentation sections.

sphinx-gallery could also be improved to use the title of the example (or even the title of the matplotlib plot object when available) as an alt text for the image. I will open an issue.

ogrisel commented:
Also, on small screens, when the burger menu is active, one could get rid of the nested "more" menu and expand it inline in the burger menu itself by default (but maybe it's too complicated to do in which case it's fine as it is).

GaelVaroquaux committed:
STY: increase contrast (#11)

thomasjpfan committed:
CLN Refactor nav list

thomasjpfan committed:
ENH Adjust for lighthouse audit

thomasjpfan committed:
BUG Fix

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into scikit-learn-bootstrap

thomasjpfan committed:
TST [doc build]

thomasjpfan committed:
STY Fixes deprecated box

thomasjpfan committed:
STY Adjust spacing

thomasjpfan committed:
STY versionchanged styping

thomasjpfan committed:
ENH Better TOC nav [doc build]

thomasjpfan committed:
STY Update styling [doc build]

thomasjpfan committed:
STY Updates

thomasjpfan committed:
TST [doc build]

thomasjpfan committed:
STY Slightly more shadow [doc build]

thomasjpfan committed:
STY Improves word-break on sidebar [doc build]

thomasjpfan committed:
STY Better wrapping [doc build]

thomasjpfan committed:
STY Adjust API classes style [doc build]

jnothman commented:
Are there things you want reviewed, or should we just check a little and
merge?

thomasjpfan commented on doc/themes/scikit-learn-modern/layout.html:
```diff
@@ -0,0 +1,128 @@
+            Please <a class="font-weight-bold" href="{{ pathto('about').replace('#', '') }}#citing-scikit-learn"><string>cite us</string></a> if you use the software.
+          </p>
+        </div>
+          {%- if meta and meta['globalsidebartoc']|tobool %}
```
This was added to be able to set the sidebar TOC by adding:

```
:globalsidebartoc: True
```

to a rst file. Are we okay with something like this?

GaelVaroquaux commented:
> :globalsidebartoc: True

> to a rst file. Are we okay with something like this?

I suspect that just about nobody will use it (it's impressive how people
do not know most rst / sphinx features). That said, if you add it to the
relevant places in this PR, it's useful.

thomasjpfan committed:
STY Minor for versionchanged

thomasjpfan commented:
> I suspect that just about nobody will use it (it's impressive how people
do not know most rst / sphinx features). That said, if you add it to the
relevant places in this PR, it's useful.

It is used in places where the local toc does not show much. For example the sidebar of the user guide:

https://thomasjpfan.github.io/scikit-learn-website/user_guide.html

https://scikit-learn.org/stable/user_guide.html

jnothman commented:
Is it really about global toc, or about parent toc? For example pages,
parent toc is appropriate. For example:
https://thomasjpfan.github.io/scikit-learn-website/auto_examples/datasets/plot_digits_last_image.html#sphx-glr-auto-examples-datasets-plot-digits-last-image-py

GaelVaroquaux commented:
> It is used in places where the local toc does not show much. For example the sidebar of the user guide:

That's good with me. Did you add a comment in the rst there? (sorry, I
should look it up, I'm lazy).

thomasjpfan committed:
STY CSS magic for api docs [doc build]

thomasjpfan committed:
BLD Update mathjax [doc build]

thomasjpfan committed:
DOC Adds comments about global toc [doc build]

thomasjpfan commented:
Before I forget to say: @ogrisel concerns with lighthouse were addressed

> Is it really about global toc, or about parent toc?

@jnothman It is global. It would be awesome to get the parent TOC. Currently the sphinx API only has access to the global and local toc in a template:

https://www.sphinx-doc.org/en/master/templating.html#toc

There are a few other places where the sidebar can be improved, such as:

1. [The class API doc sidebar](https://thomasjpfan.github.io/scikit-learn-website/modules/generated/sklearn.linear_model.Perceptron.html#sklearn.linear_model.Perceptron) can contain the other classes in the `linear_model` module.
2. As you said, the examples could at least include the other examples in the same section.

Both of these most likely require something clever with sphinx, (which can be done in another PR if we want it).

> That's good with me. Did you add a comment in the rst there? (sorry, I
should look it up, I'm lazy).

There is one now. ;)

thomasjpfan committed:
DOC Adds comments about global toc [doc build]

ogrisel reviewed (approved):
I agree that the class API sidebar could be improved but I believe that this PR is already a net improvement.

I wasn't so sure about the the colored gradient as the background of the top-level banner but as long as the readability is good enough (which is the case now I think), I welcome the overall style refresh :)

Also I really like the inline example plots.

ogrisel commented on doc/templates/index.html:
```diff
@@ -0,0 +1,256 @@
+        </div>
+        <div class="overflow-hidden mx-2 text-center flex-fill">
+          <a href="auto_examples/classification/plot_classifier_comparison.html"  aria-label="Classification">
+          <img src="_images/sphx_glr_plot_classifier_comparison_001_carousel.png" class="sk-index-img" style="width: unset;" alt="Classification">
```
Actually for the alt text it would be better to be more informative about what the image is actually about (e.g. the title of the specific example rather than the title of the section).

GaelVaroquaux commented:
With regards to the state of this PR, I'm still +1 for merge.

However, if it's possible to get back the "sticky" top bar on mobile
phone, it would be great. It works on large screens.

thomasjpfan commented:
> However, if it's possible to get back the "sticky" top bar on mobile
phone, it would be great. It works on large screens.

It had some scrolling bugs when the navbar was expanded. (Especially now that the "More" tab is expanded by default on mobile.)

I'll see what I can do about it.

thomasjpfan committed:
DOC Better alt names

GaelVaroquaux commented:
> I'll see what I can do about it.

If it's too hard, just give up and we merge.

thomasjpfan committed:
ENH Adds fixed navbar in mobile [doc build]

thomasjpfan commented:
> If it's too hard, just give up and we merge.

Fixed navbar added back to mobile.

thomasjpfan committed:
MAINT Remove unused css [doc build]

ogrisel commented:
@jnothman any further comments? merge?

GaelVaroquaux commented:
One last thing ;).

I think that it is useful to add a max-width: 100% rule on the images of
the body. On small devices, these can cause horizontal scroll. Horizontal
scroll is particularly a problem because the burger menu of the top bar
becomes hidden.

jnothman reviewed (commented):
Otherwise, good to merge!!

Wow, @thomasjpfan

thomasjpfan commented on doc/themes/scikit-learn-modern/javascript.html:
```diff
@@ -0,0 +1,160 @@
+	});
+
+{%- if pagename != 'index' and pagename != 'documentation' %}
+  /*** Had navbar when scrolling down ***/
```
It meant to "Hide"

thomasjpfan commented on doc/themes/scikit-learn-modern/theme.conf:
```diff
@@ -0,0 +1,8 @@
+stylesheet = css/theme.css
+
+[options]
+google_analytics = false
```
This sets the default. I can change it to true by default.

thomasjpfan replied:
This was changed to true.

thomasjpfan commented on doc/themes/scikit-learn-modern/javascript.html:
```diff
@@ -0,0 +1,160 @@
+{% endif %}
+
+<script>
+$(document).ready(function() {
```
This reduces the number of HTTP requests.

thomasjpfan replied:
The total number of bytes the javascript.html is small (in bytes), so I thought it was worth it to inline the javascript.

thomasjpfan commented on doc/templates/index.html:
```diff
@@ -0,0 +1,256 @@
+        <a class="btn btn-warning btn-big sk-donate-btn mb-1" onclick="document.getElementById('paypal-form').submit(); ">Help us, <strong>donate!</strong></a>
+        <a class="btn btn-warning btn-big mb-1" href="about.html#citing-scikit-learn"><strong>Cite us!</strong></a>
+        <p class="mt-2">
+            <a href="about.html#funding">Read more about donations</a>
```
Poor as in its too small?

thomasjpfan committed:
CLN Address comments

thomasjpfan commented on doc/conf.py:
```diff
@@ -314,8 +309,8 @@ def make_carousel_thumbs(app, exception):
     # to hide/show the prompt in code examples:
-    app.add_javascript('js/copybutton.js')
-    app.add_javascript('js/extra.js')
+    # app.add_javascript('js/copybutton.js')
```
Removed.

thomasjpfan committed:
TST [doc build]

GaelVaroquaux committed:
Avoid images overflowing horizontally on small screens (#12)

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into scikit-learn-bootstrap

thomasjpfan committed:
TST [doc build]

GaelVaroquaux commented:
LGTM.

Merge? @jnothman : you had comments. I'll leave you some time to voice any other comments if you have some. If I don't hear from you, I'll merge.

thomasjpfan committed:
DOC Adds getting started [doc build]

jnothman commented:
Awesome work, @thomasjpfan!

amueller commented:
Sorry I didn't follow the discussion.
If I click on "install" the sidebar has a toc of install, and similarly for examples and API. But if I click on User Guide, the sidebar has a toc for the full page that includes the API and examples. That's a bit confusing. Here the sidebar doesn't correspond to the things on the left (which kind of makes sense since you don't want to show the same thing twice), but I find it confusing.

amueller commented:
also, the "documentation" page now seems a bit useless, right?

amueller commented:
also weird line-wrap:
![image](https://user-images.githubusercontent.com/449558/65441244-12e2eb80-ddf8-11e9-8ff9-92a21979c2b7.png)

amueller commented:
Also, I would maybe remove the testimonials. They are mostly confusing if anything. Linking to the sklearn consortium or the google scholar citations would be more informative, I think.

thomasjpfan commented:
> If I click on "install" the sidebar has a toc of install, and similarly for examples and API. But if I click on User Guide, the sidebar has a toc for the full page that includes the API and examples. That's a bit confusing. Here the sidebar doesn't correspond to the things on the left (which kind of makes sense since you don't want to show the same thing twice), but I find it confusing.

The sidebar changes based on context. Before the sidebar was just useless: https://scikit-learn.org/stable/user_guide.html

> also, the "documentation" page now seems a bit useless, right?

It is less useful now. It still gives a little context to each link.

amueller commented:
There has been a discussion on how to structure numpy, scipy and pandas websites here:
https://github.com/numpy/numpy.org/issues/43

It might make sense to join the discussion and/or follow suite.

amueller commented:
> It is less useful now. It still gives a little context to each link.

well but when/how would a user end up there? like if they don't know what of the other links to click? seems like a strange path to take.

thomasjpfan commented:
> well but when/**how** would a user end up there

They can end up there from the landing page if they click "Documentation":

![Screen Shot 2019-09-23 at 4 41 24 PM](https://user-images.githubusercontent.com/5402633/65460976-0f178f00-de21-11e9-9bfa-dbe196b717cc.png)

> well but **when**/how would a user end up there? like if they don't know what of the other links to click? seems like a strange path to take.

It kind of serves as a global index. I am okay with removing it.

amueller commented:
Ah, I hadn't seen that. I'm not opposed to keeping it but it seems less useful now.

ogrisel commented:
Same feeling.

thomasjpfan commented:
One of the ways we are using the documentation page is to outline all the versions in the [all versions page](https://scikit-learn.org/dev/versions.html) (It shows the version number on the top of the page)

lesteve commented on doc/themes/scikit-learn-modern/static/css/theme.css:
```diff
@@ -0,0 +1,1096 @@
+
+
+@media screen and (min-width: 1540px) {
+  .sphx-glr-download-link-note {
```
Great stuff! I just noticed that the note about the binder link is floated right, I wish I had the web skills to think of this in the first place and then execute it as well.
![image](https://user-images.githubusercontent.com/1680079/67147508-cc1cd000-f295-11e9-9ed6-36089903c934.png)

GaelVaroquaux commented:
> Great stuff! I just noticed that the note about the binder link is floated
> right, I wish I had the web skills to think of this in the first place and then
> execute it as well.

Should we consider backporting this to sphinx-gallery?

lesteve commented:
> Should we consider backporting this to sphinx-gallery?

Sounds like a good idea indeed.
